Add option to save the BuildSourcesEphemeral overlay.

behrmann opened the pull request:
This systematises a trick Daan showed me a while back and that I've been using most everywhere since. Having it in mkosi itself makes it easier to use for everybody.

DaanDeMeyer reviewed (commented):
So I'm not too sure about adding official support for my hack, how about we document it somewhere instead?

behrmann commented:
I'm also happy to call the option `voidwarranty-buildcache` instead.

behrmann commented:
Rebased and dropped the dont merge label, since it worked locally. I cherry-picked 62582f3d0d545c41264dfeed3f5880580b8dc29a from #3402 for now to make CI run.

DaanDeMeyer reviewed (changes requested):
Shouldn't we limit this to only when running build scripts and do regular ephemeral for the other scripts?

behrmann commented:
Good catch, I thought it was there, that's why it's mentioned in the man page. I must have lost this in a rebase.

DaanDeMeyer commented on mkosi/mounts.py:
```diff
@@ -56,20 +56,30 @@ def mount_overlay(
             if ephemeral:
-                upperdir = Path(stack.enter_context(tempfile.TemporaryDirectory(prefix="volatile-overlay")))
-                os.chmod(upperdir, src.stat().st_mode)
+                if ephemeral == BuildSourcesEphemeral.buildcache and config.build_dir is not None:
```
I would fail if this is specified but there's no build directory configured

DaanDeMeyer commented on mkosi/mounts.py:
```diff
@@ -56,20 +57,36 @@ def mount_overlay(
+                            "BuildSourcesEphemeral=buildcache was configured, but no build directory exists.",  # noqa: E501
+                            hint="Configure BuildDirectory= or create mkosi.builddir.",
+                        )
+                    assert config.build_dir
```
This shouldn't be needed anymore no?
